spec/unit/action/deploy/environment_spec.rb

```diff
@@ -275,13 +284,17 @@
         it "only logs about purging environment" do
           expect(subject.logger).to receive(:debug).with(/purging unmanaged content for environment/i)
           expect(subject.logger).to_not receive(:debug).with(/purging unmanaged environments for deployment/i)
-          expect(subject.logger).to_not receive(:debug).with(/purging unmanaged puppetfile content/i)
```

Comment:
So whenever I tried to implement a similar "wrap_original" expectation on "visit_environment" like above to test for the logger output of "purging unmanaged puppetfile content" the remaining expectations wouldn't fire. I think that's a testing issue though I'm unclear on what the root cause is. I wanted to get other's opinion on it. This test is testing less, though I don't know if it's any worse.

Reply:
Weird that it works in one case but not the other... I feel like I ran into something weird like this recently with some work I was doing on the CA CLI tests, but I can't remember if/how I fixed it... I'll think about it.

Reply:
So... it doesn't actually look like you need the `and_wrap_original` at all. From my testing, `env.logger` is the same as `subject.logger`, so you can just add both expectations to `subject.logger` directly. If I do that, the test passes. Works for both for me.

Reply:
I don't know why using `and_wrap_original` doesn't work though 🤷‍♀️

Reply:
According to pry, that `env` object doesn't have a `logger` method (or instance var, for that matter). So I _think_ this block is actually throwing and the exception is being squashed, but it's messing with other stuff.

It does have a `Puppetfile` object that has a logger though, if I put the expectation on that object, the tests pass.

Reply:
Ah ha! So it looks like the logger functionality is included in all the child Environment objects, rather than the Environment::Base class. This deployment uses Environment::Mock which doesn't include the logging framework.

Including the R10K::Logging module into the Environment::Mock class causes _a different_ test to fail.... So I think I should move the inclusion of the Logging framework into the Environment::Base and then audit the usages of loggers & the mock environment...

Reply:
That makes sense to me. Yeah, it took me a bit to track down exactly where `logger` was coming from. I expected it to be part of the `Base` environment...

Reply:
Okay, I moved the logging inclusion for the Environments and while I was at it for the Sources too (since they also have a Mock version that doesn't include a logger). There was also an issue with the test data not being passed down from Source::Mock to Environment::Mock correctly. I think that and the silently failing logger calls were the two issues conspiring against my tests before.